bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -556,11 +556,36 @@ impl AuthenticationService {
     ) -> Result<VerifiedClientMetadata, AuthenticationError> {
         let redirect_uri = Url::parse(&configuration.redirect_uri)
             .map_err(|_| AuthenticationError::OidcCallbackUrlInvalid)?;
-        let client_name = Some(Localized::new(configuration.client_name.to_owned(), []));
-        let client_uri = Url::parse(&configuration.client_uri).ok().map(|u| Localized::new(u, []));
-        let logo_uri = Url::parse(&configuration.logo_uri).ok().map(|u| Localized::new(u, []));
-        let policy_uri = Url::parse(&configuration.policy_uri).ok().map(|u| Localized::new(u, []));
-        let tos_uri = Url::parse(&configuration.tos_uri).ok().map(|u| Localized::new(u, []));
+        let client_name =
+            configuration.client_name.as_ref().map(|n| Localized::new(n.to_owned(), []));
+        let client_uri = configuration
+            .client_uri
+            .as_deref()
+            .map(Url::parse)
+            .transpose()
+            .map_err(|_| AuthenticationError::OidcMetadataInvalid)?
+            .map(|u| Localized::new(u, []));
+        let logo_uri = configuration
+            .logo_uri
+            .as_deref()
+            .map(Url::parse)
+            .transpose()
+            .map_err(|_| AuthenticationError::OidcMetadataInvalid)?
+            .map(|u| Localized::new(u, []));
```

Comment:
Might be easier to read?

```suggestion
            .map(|uri| {
                Ok(Localized::new(
                    Url::parse(uri).map_err(|_| AuthenticationError::OidcMetadataInvalid)?,
                    [],
                ))
            })
            .transpose()?;
```

Reply:
Updated.

Reply:
Also, I tried doing an impl on `Option<&str>` to reduce duplicated code, but I don't think that's allowed in Rust?

Reply:
Updated with an extension trait 😃